File: internal/dashboard/contracts/harness.go
package contracts

// Pipeline stages per docs/specs/web-portal/harness-pipeline-data-model.md.

var PipelineStageNames = []string{
	"Plan",
	"Delegate",
	"Execute",
	"Propose",
	"Court Review",
	"Apply",
}

// PlanStatus values for harness plans.
const (
	PlanStatusActive    = "active"
	PlanStatusCompleted = "completed"
	PlanStatusCancelled = "cancelled"
)

// StageStatus values for pipeline stages.
const (
	StagePending    = "pending"
	StageInProgress = "in_progress"
	StageCompleted  = "completed"
	StageFailed     = "failed"
)

// Plan is the high-level harness goal for a channel.
type Plan struct {
	PlanID    string         `json:"plan_id"`
	ChannelID string         `json:"channel_id"`
	Goal      string         `json:"goal"`
	CreatedAt string         `json:"created_at"`
	Status    string         `json:"status"`
	Stages    []StageStatus  `json:"stages"`
}

// StageStatus is one pipeline stage with status.
type StageStatus struct {
	Name   string `json:"name"`
	Status string `json:"status"`
}

// NarrowTask is a decomposed unit of work for a specialist persona.
// agent_instance_id must never be serialized to browser-bound JSON.
type NarrowTask struct {
	TaskID       string `json:"task_id"`
	PlanID       string `json:"plan_id"`
	AgentPersona string `json:"agent_persona"`
	Scope        string `json:"scope"`
	Status       string `json:"status"`
	CurrentStage string `json:"current_stage"`
	Progress     int    `json:"progress"`
	LastUpdate   string `json:"last_update"`
}

// HarnessState aggregates plan + tasks for channel/dashboard views.
type HarnessState struct {
	Plan  *Plan        `json:"plan,omitempty"`
	Tasks []NarrowTask `json:"tasks"`
}

// DefaultStages returns pending pipeline stages for a new plan.
func DefaultStages() []StageStatus {
	out := make([]StageStatus, len(PipelineStageNames))
	for i, name := range PipelineStageNames {
		status := StagePending
		if i == 0 {
			status = StageInProgress
		}
		out[i] = StageStatus{Name: name, Status: status}
	}
	return out
}

// HarnessPlanCreated event payload.
type HarnessPlanCreated struct {
	Type      string        `json:"type"`
	PlanID    string        `json:"plan_id"`
	ChannelID string        `json:"channel_id"`
	Goal      string        `json:"goal"`
	Stages    []StageStatus `json:"stages"`
}

// HarnessTaskAssigned event payload.
type HarnessTaskAssigned struct {
	Type          string `json:"type"`
	TaskID        string `json:"task_id"`
	PlanID        string `json:"plan_id"`
	AgentPersona  string `json:"agent_persona"`
	Scope         string `json:"scope"`
	CurrentStage  string `json:"current_stage"`
}

// HarnessTaskProgress event payload.
type HarnessTaskProgress struct {
	Type         string `json:"type"`
	TaskID       string `json:"task_id"`
	Progress     int    `json:"progress"`
	CurrentStage string `json:"current_stage"`
	Summary      string `json:"summary"`
}

// HarnessStageTransition event payload.
type HarnessStageTransition struct {
	Type           string   `json:"type"`
	PlanID         string   `json:"plan_id"`
	Stage          string   `json:"stage"`
	Status         string   `json:"status"`
	RelatedTaskIDs []string `json:"related_task_ids"`
}

// HarnessProposalCreated event payload.
type HarnessProposalCreated struct {
	Type       string `json:"type"`
	PlanID     string `json:"plan_id"`
	TaskID     string `json:"task_id"`
	ProposalID string `json:"proposal_id"`
	Stage      string `json:"stage"`
}

File: web-portal/src/lib/harness.ts
import {
  EVENT,
  HarnessState,
  PIPELINE_STAGES,
  StageStatus,
} from '@/contracts';

export function defaultStages(): StageStatus[] {
  return PIPELINE_STAGES.map((name, i) => ({
    name,
    status: i === 0 ? 'in_progress' : 'pending',
  }));
}

export function applyHarnessEvent(
  state: HarnessState,
  event: Record<string, unknown>,
): HarnessState {
  if (!event?.type) return state;
  const next: HarnessState = structuredClone(state || { plan: null, tasks: [] });

  switch (event.type) {
    case EVENT.harnessPlanCreated:
      next.plan = {
        plan_id: String(event.plan_id ?? ''),
        channel_id: String(event.channel_id ?? ''),
        goal: String(event.goal ?? ''),
        stages: (event.stages as StageStatus[]) || defaultStages(),
        status: 'active',
      };
      break;
    case EVENT.harnessTaskAssigned:
      next.tasks = next.tasks || [];
      next.tasks.push({
        task_id: String(event.task_id ?? ''),
        plan_id: String(event.plan_id ?? ''),
        agent_persona: String(event.agent_persona ?? 'agent'),
        scope: String(event.scope ?? ''),
        current_stage: String(event.current_stage ?? 'Execute'),
        status: 'active',
        progress: 0,
      });
      break;
    case EVENT.harnessTaskProgress:
      next.tasks = (next.tasks || []).map((t) =>
        t.task_id === event.task_id
          ? {
              ...t,
              progress: Number(event.progress ?? t.progress),
              current_stage: String(event.current_stage ?? t.current_stage),
              summary: String(event.summary ?? t.summary ?? ''),
            }
          : t,
      );
      break;
    case EVENT.harnessStageTransition:
      if (next.plan?.stages) {
        next.plan.stages = next.plan.stages.map((s) =>
          s.name === event.stage
            ? { ...s, status: (event.status as StageStatus['status']) || s.status }
            : s,
        );
      }
      break;
    default:
      break;
  }
  return next;
}

export function slugStage(name: string): string {
  return String(name || '').toLowerCase().replace(/\s+/g, '-');
}